File: tools/p0_preflight.R
#!/usr/bin/env Rscript
# Read-only inventory of the supported native/R toolchain, never installation.
source(file.path(dirname(sub('^--file=','',grep('^--file=',commandArgs(),value=TRUE)[1])),'lib_helpers.R'))
flags<-list('g++'='--version',make='--version',Rscript='--version',STAR='--version','hisat2-align-s'='--version','hisat2-build-s'='--version',featureCounts='-v',samtools='--version','fasterq-dump'='--version',sbatch='--version')
report<-list(platform=as.list(Sys.info()),R=R.version.string,tools=list())
for(name in names(flags)){path<-Sys.which(name);report$tools[[name]]<-if(!nzchar(path))list(path=NULL)else c(list(path=unname(path)),tryCatch(run(c(path,flags[[name]]),timeout=30,check=FALSE),error=function(e)list(error=conditionMessage(e))))}
report$r_packages<-setNames(lapply(c('DESeq2','apeglm','BiocParallel','pheatmap','jsonlite'),function(name)if(requireNamespace(name,quietly=TRUE))as.character(packageVersion(name))else 'MISSING'),c('DESeq2','apeglm','BiocParallel','pheatmap','jsonlite'))
cat(jsonlite::toJSON(report,auto_unbox=TRUE,pretty=TRUE,null='null'),'\n')

File: tools/p0_preflight.py
#!/usr/bin/env python3
"""Read-only local tool inventory; never installs tools or submits jobs."""
import importlib.util
import json
import platform
import shutil
import subprocess
import sys


def probe(argv):
    try:
        result = subprocess.run(argv, capture_output=True, text=True, timeout=30)
        return {"exit_code": result.returncode,
                "output": (result.stdout + result.stderr).strip()}
    except (OSError, subprocess.TimeoutExpired) as exc:
        return {"error": str(exc)}


def main():
    flags = {"g++": "--version", "make": "--version", "Rscript": "--version",
             "STAR": "--version", "hisat2": "--version", "featureCounts": "-v",
             "samtools": "--version", "snakemake": "--version",
             "fasterq-dump": "--version", "sbatch": "--version", "node": "--version"}
    report = {"platform": platform.platform(), "python": sys.version,
              "python_executable": sys.executable, "tools": {}}
    for name, flag in flags.items():
        path = shutil.which(name)
        report["tools"][name] = {"path": path, **(probe([path, flag]) if path else {})}
    report["python_modules"] = {
        name: importlib.util.find_spec(name) is not None
        for name in ("pandas", "biomart", "loguru", "Bio", "snakemake")}
    if shutil.which("Rscript"):
        report["r_packages"] = probe([
            "Rscript", "--vanilla", "-e",
            'for (p in c("DESeq2","apeglm","GEOquery","optparse","tidyverse",'
            '"EnhancedVolcano","RColorBrewer","pheatmap","ggrepel",'
            '"org.Hs.eg.db","BiocParallel")) cat(p, '
            'if (requireNamespace(p, quietly=TRUE)) as.character(packageVersion(p)) '
            'else "MISSING", "\\n")'])
    print(json.dumps(report, indent=2))


if __name__ == "__main__":
    main()
